### ISSUE TITLE
Support node_roles in deployment commands

### ISSUE THREAD

**Issue by marclop** [elastic/ecctl#451]
## Overview

With the introduction of `node_roles` and the `node_type` object being marked as deprecated going forward in the stack 7.11+, we nee a good strategy to update the code where `node_type` is being used and use `node_roles` if 7.11+.

A good discussion point is to use the deployment template instead of generating our own payload through the CLI flags like we are doing now, and use the CLI flags to set overrides.

**Comment by kuisathaverat** [comment]
I have followed your suggestion to create a new plan from the template deployed with the following command I've grad the base of the plan

`ecl deployment template show --template-id="aws-hot-warm-v3" --output=json  --host https://admin.staging.foundit.no | jq '.deployment_template| map(del(.resources.elasticsearch[0].plan.elasticsearch.curation))'`

then I added the `deployment_template` and the `version` to complete my plan

```
{
  "resources": {
...
    "elasticsearch": [
      {
        "plan": {
...


          "deployment_template": {
            "id": "aws-hot-warm-v3"
          },
          "elasticsearch": {
            "version": "7.11.0"
          }

...
       }
...

```

And this is the final plan

```
{
  "name": "oblt-test-ecl",
  "resources": {
    "apm": [
      {
        "elasticsearch_cluster_ref_id": "elasticsearch-main",
        "plan": {
          "apm": {},
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.apm.r5d",
              "size": {
                "resource": "memory",
                "value": 512
              },
              "zone_count": 1
            }
          ]
        },
        "ref_id": "apm-main",
        "region": "us-east-1"
      }
    ],
    "appsearch": null,
    "elasticsearch": [
      {
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.coordinating.m5d",
              "node_type": {
                "data": false,
                "ingest": true,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "hot"
                }
              },
              "instance_configuration_id": "aws.data.highio.i3",
              "node_type": {
                "data": true,
                "ingest": true,
                "master": true
              },
              "size": {
                "resource": "memory",
                "value": 4096
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "warm"
                }
              },
              "instance_configuration_id": "aws.data.highstorage.i3en",
              "node_type": {
                "data": true,
                "ingest": false,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "cold"
                }
              },
              "instance_configuration_id": "aws.data.highstorage.i3en",
              "node_type": {
                "data": true,
                "ingest": false,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 1
            },
            {
              "instance_configuration_id": "aws.master.r5d",
              "node_type": {
                "data": false,
                "ingest": false,
                "master": true
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 3
            },
            {
              "instance_configuration_id": "aws.ml.m5d",
              "node_type": {
                "data": false,
                "ingest": false,
                "master": false,
                "ml": true
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 1
            }
          ],
          "deployment_template": {
            "id": "aws-hot-warm-v3"
          },
          "elasticsearch": {
            "version": "7.11.0"
          }
        },
        "ref_id": "elasticsearch-main",
        "region": "us-east-1",
        "settings": {
          "curation": {
            "specs": [
              {
                "index_pattern": "logstash-*",
                "trigger_interval_seconds": 86400
              },
              {
                "index_pattern": "filebeat-*",
                "trigger_interval_seconds": 86400
              },
              {
                "index_pattern": "metricbeat-*",
                "trigger_interval_seconds": 86400
              }
            ]
          },
          "dedicated_masters_threshold": 6
        }
      }
    ],
    "enterprise_search": [
      {
        "elasticsearch_cluster_ref_id": "elasticsearch-main",
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.enterprisesearch.m5d",
              "node_type": {
                "appserver": true,
                "connector": true,
                "worker": true
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 2
            }
          ],
          "enterprise_search": {}
        },
        "ref_id": "enterprise_search-main",
        "region": "us-east-1"
      }
    ],
    "kibana": [
      {
        "elasticsearch_cluster_ref_id": "elasticsearch-main",
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.kibana.r5d",
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 1
            }
          ],
          "kibana": {}
        },
        "ref_id": "kibana-main",
        "region": "us-east-1"
      }
    ]
  }
}

```

```
ecl deployment create  --file elasticsearch-plan-update.json  --host https://admin.staging.foundit.no --track

The deployment creation returned with an error. Use the displayed request ID to recreate the deployment resources
Request ID: f7swggviidaijx9zuvmco8bept7rg8aynmj8br55lefomyzwpmwzam8960t8aaa9
api error: 2 errors occurred:
	* clusters.cluster_invalid_plan: Cluster must contain at least a master topology element and a data topology element. 'master' node type exists in more than one topology element (resources.elasticsearch[0].cluster_topology)
	* deployments.unrecognized_topology_elements: Deployment plan definition includes topology element not supported by template [Hot-Warm Architecture]. Instance configuration: [Some(aws.data.highstorage.i3en)], node types: [Some(ElasticsearchNodeType(Some(false),Some(true),Some(false),None))] (resources.elasticsearch[0])

```

then I have removed the Cold node because it fails to apply the resources, I dunno why, the template is correct

```
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "cold"
                }
              },
              "instance_configuration_id": "aws.data.highstorage.i3en",
              "node_type": {
                "data": true,
                "ingest": false,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 1
            },
```

for some reason does not get the master and data node

```
❯ ecl deployment create  --file elasticsearch-plan-update.json  --host https://admin.staging.foundit.no --track

The deployment creation returned with an error. Use the displayed request ID to recreate the deployment resources
Request ID: 3m4q79qhagmwai4i52soa8hy1ilkbhd2i1d19pxvpotzw0d7xlmmycxfdyzv3nur
api error: 1 error occurred:
	* clusters.cluster_invalid_plan: Cluster must contain at least a master topology element and a data topology element. 'master' node type exists in more than one topology element (resources.elasticsearch[0].cluster_topology)
```

node_role is documented as an array of strings at https://www.elastic.co/guide/en/cloud/current/Deployment_-_CRUD.html#ec_request_example_13 but I did not found examples in the documentation, there is an issue to document it https://github.com/elastic/cloud/issues/73820 Digging in code I have found https://github.com/elastic/cloud/issues/74378 that has an example with some values valid.

```
{
  "name": "oblt-test-ecl",
  "resources": {
    "apm": [
      {
        "elasticsearch_cluster_ref_id": "elasticsearch-main",
        "plan": {
          "apm": {},
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.apm.r5d",
              "size": {
                "resource": "memory",
                "value": 512
              },
              "zone_count": 1
            }
          ]
        },
        "ref_id": "apm-main",
        "region": "us-east-1"
      }
    ],
    "appsearch": null,
    "elasticsearch": [
      {
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.coordinating.m5d",
              "node_roles": [
                "ingest",
                "remote_cluster_client"
              ],
              "node_type": {
                "data": false,
                "ingest": true,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "hot"
                }
              },
              "instance_configuration_id": "aws.data.highio.i3",
              "node_roles": [
                "master",
                "ingest",
                "remote_cluster_client",
                "data_hot",
                "transform",
                "data_content"
              ],
              "node_type": {
                "data": true,
                "ingest": true,
                "master": true
              },
              "size": {
                "resource": "memory",
                "value": 4096
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "warm"
                }
              },
              "instance_configuration_id": "aws.data.highstorage.i3en",
              "node_roles": [
                "data_warm",
                "remote_cluster_client"
              ],
              "node_type": {
                "data": true,
                "ingest": false,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 2
            },
            {
              "instance_configuration_id": "aws.master.r5d",
              "node_roles": [
                "master"
              ],
              "node_type": {
                "data": false,
                "ingest": false,
                "master": true
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 3
            },
            {
              "instance_configuration_id": "aws.ml.m5d",
              "node_roles": [
                "ml",
                "remote_cluster_client"
              ],
              "node_type": {
                "data": false,
                "ingest": false,
                "master": false,
                "ml": true
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 1
            }
          ],
          "deployment_template": {
            "id": "aws-hot-warm-v3"
          },
          "elasticsearch": {
            "version": "7.11.0"
          }
        },
        "ref_id": "elasticsearch-main",
        "region": "us-east-1",
        "settings": {
          "curation": {
            "specs": [
              {
                "index_pattern": "logstash-*",
                "trigger_interval_seconds": 86400
              },
              {
                "index_pattern": "filebeat-*",
                "trigger_interval_seconds": 86400
              },
              {
                "index_pattern": "metricbeat-*",
                "trigger_interval_seconds": 86400
              }
            ]
          },
          "dedicated_masters_threshold": 6
        }
      }
    ],
    "enterprise_search": [
      {
        "elasticsearch_cluster_ref_id": "elasticsearch-main",
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.enterprisesearch.m5d",
              "node_type": {
                "appserver": true,
                "connector": true,
                "worker": true
              },
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 2
            }
          ],
          "enterprise_search": {}
        },
        "ref_id": "enterprise_search-main",
        "region": "us-east-1"
      }
    ],
    "kibana": [
      {
        "elasticsearch_cluster_ref_id": "elasticsearch-main",
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.kibana.r5d",
              "size": {
                "resource": "memory",
                "value": 2048
              },
              "zone_count": 1
            }
          ],
          "kibana": {}
        },
        "ref_id": "kibana-main",
        "region": "us-east-1"
      }
    ]
  }
}

```

no luck with that template too

```
❯ ecl deployment create  --file elasticsearch-plan-update.json  --host https://admin.staging.foundit.no --track
The deployment creation returned with an error. Use the displayed request ID to recreate the deployment resources
Request ID: 3m4q79qhagmwai4i52soa8hy1ilkbhd2i1d19pxvpotzw0d7xlmmycxfdyzv3nur
api error: 1 error occurred:
	* clusters.cluster_invalid_plan: Cluster must contain at least a master topology element and a data topology element. 'master' node type exists in more than one topology element (resources.elasticsearch[0].cluster_topology)
```

so I have no clue how to set a node as master or data `node_type` and `node_role` are ignored

**Comment by kuisathaverat** [comment]
I have changed to use `aws-hot-warm-v2` that template works 

```
{
  "name": "oblt-test-ecl",
  "resources": {
    "apm": [
      {
        "elasticsearch_cluster_ref_id": "es-ref-id",
        "plan": {
          "apm": {},
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.apm.r5d",
              "size": {
                "resource": "memory",
                "value": 512
              },
              "zone_count": 1
            }
          ]
        },
        "ref_id": "apm-ref-id",
        "region": "us-east-1"
      }
    ],
    "appsearch": null,
    "elasticsearch": [
      {
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.coordinating.m5d",
              "node_type": {
                "data": false,
                "ingest": true,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 0
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "hot"
                }
              },
              "instance_configuration_id": "aws.data.highio.i3",
              "node_type": {
                "data": true,
                "ingest": true,
                "master": true
              },
              "size": {
                "resource": "memory",
                "value": 4096
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "warm"
                }
              },
              "instance_configuration_id": "aws.data.highstorage.d2",
              "node_type": {
                "data": true,
                "ingest": true,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 4096
              },
              "zone_count": 2
            },
            {
              "elasticsearch": {
                "node_attributes": {
                  "data": "cold"
                }
              },
              "instance_configuration_id": "aws.data.highstorage.d2",
              "node_type": {
                "data": true,
                "ingest": false,
                "master": false
              },
              "size": {
                "resource": "memory",
                "value": 0
              },
              "zone_count": 1
            },
            {
              "instance_configuration_id": "aws.master.r5d",
              "node_type": {
                "data": false,
                "ingest": false,
                "master": true
              },
              "size": {
                "resource": "memory",
                "value": 0
              },
              "zone_count": 3
            },
            {
              "instance_configuration_id": "aws.ml.m5d",
              "node_type": {
                "data": false,
                "ingest": false,
                "master": false,
                "ml": true
              },
              "size": {
                "resource": "memory",
                "value": 0
              },
              "zone_count": 1
            }
          ],
          "deployment_template": {
            "id": "aws-hot-warm-v2"
          },
          "elasticsearch": {
            "version": "7.11.0"
          }
        },
        "ref_id": "es-ref-id",
        "region": "us-east-1",
        "settings": {
          "curation": {
            "specs": [
              {
                "index_pattern": "logstash-*",
                "trigger_interval_seconds": 86400
              },
              {
                "index_pattern": "filebeat-*",
                "trigger_interval_seconds": 86400
              },
              {
                "index_pattern": "metricbeat-*",
                "trigger_interval_seconds": 86400
              }
            ]
          },
          "dedicated_masters_threshold": 6
        }
      }
    ],
    "enterprise_search": [
      {
        "elasticsearch_cluster_ref_id": "es-ref-id",
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.enterprisesearch.m5d",
              "node_type": {
                "appserver": true,
                "connector": true,
                "worker": true
              },
              "size": {
                "resource": "memory",
                "value": 0
              },
              "zone_count": 2
            }
          ],
          "enterprise_search": {}
        },
        "ref_id": "enterprise_search-ref-id",
        "region": "us-east-1"
      }
    ],
    "kibana": [
      {
        "elasticsearch_cluster_ref_id": "es-ref-id",
        "plan": {
          "cluster_topology": [
            {
              "instance_configuration_id": "aws.kibana.r5d",
              "size": {
                "resource": "memory",
                "value": 1024
              },
              "zone_count": 1
            }
          ],
          "kibana": {}
        },
        "ref_id": "kibana-ref-id",
        "region": "us-east-1"
      }
    ]
  }
}

```

**Comment by marclop** [comment]
Fixed by https://github.com/elastic/ecctl/pull/460. Will be published in the upcoming `v1.3.0`.
